jax/experimental/ode.py

```diff
@@ -1,354 +1,368 @@
+
+  Args:
+    ofunc: Function `ydot = ofunc(y, t, *args)` to compute the time
+      derivative of `y`.
+    args: Additional arguments to `ofunc`.
+
+  Returns:
+    VJP function `vjp = vjp_all(g, yt, t)` where `yt = ofunc(y, t, *args)`
+    and g is used for VJP calculation. To evaluate the gradient w/ the VJP,
+    supply `g = np.ones_like(yt)`.
+  """
+  flat_args, unravel_args = ravel_pytree(args)
+  flat_func = lambda y, t, flat_args: ofunc(y, t, *unravel_args(flat_args))
+
+  @jax.jit
+  def aug_dynamics(augmented_state, t):
+    """Original system augmented with vjp_y, vjp_t and vjp_args."""
+    D = int(np.floor_divide(
+        augmented_state.shape[0] - flat_args.shape[0] - 1, 2))
+    y = augmented_state[:D]
+    adjoint = augmented_state[D:2*D]
+    dy_dt, vjpfun = jax.vjp(flat_func, y, t, flat_args)
+    return np.hstack([np.ravel(dy_dt), np.hstack(vjpfun(-adjoint))])
+
+  rev_aug_dynamics = lambda y, t: -aug_dynamics(y, -t)
+
+  @jax.jit
+  def _fori_body_fun(i, val):
+    """fori_loop function for VJP calculation."""
+    rev_yt, rev_t, rev_tarray, rev_gi, vjp_y, vjp_t0, vjp_args, time_vjp_list = val
+    this_yt = rev_yt[i, :]
+    this_t = rev_t[i]
+    this_tarray = rev_tarray[i, :]
+    this_gi = rev_gi[i, :]
+    this_gim1 = rev_gi[i-1, :]
+    D = this_yt.shape[0]
+    vjp_cur_t = np.dot(flat_func(this_yt, this_t, flat_args), this_gi)
+    vjp_t0 = vjp_t0 - vjp_cur_t
+    # Run augmented system backwards to the previous observation.
+    aug_y0 = np.hstack((this_yt, vjp_y, vjp_t0, vjp_args))
+    aug_ans = odeint(rev_aug_dynamics, (), aug_y0, this_tarray)
+    vjp_y = aug_ans[1][D:2*D] + this_gim1
+    vjp_t0 = aug_ans[1][2*D]
+    vjp_args = aug_ans[1][2*D+1:]
+    time_vjp_list = jax.ops.index_update(time_vjp_list, i, vjp_cur_t)
+    return rev_yt, rev_t, rev_tarray, rev_gi, vjp_y, vjp_t0, vjp_args, time_vjp_list
+
+  @jax.jit
+  def vjp_all(g, yt, t):
+    """Calculate the VJP g * Jac(odeint(ofunc(yt, t, *args), t)."""
+    rev_yt = yt[-1:0:-1, :]
+    rev_t = t[-1:0:-1]
+    rev_tarray = -np.array([t[-1:0:-1], t[-2::-1]]).T
+    rev_gi = g[-1:0:-1, :]
+
+    vjp_y = rev_gi[-1, :]
+    vjp_t0 = 0.
+    vjp_args = np.zeros_like(flat_args)
+    time_vjp_list = np.zeros_like(t)
+
+    result = jax.lax.fori_loop(0,
+                               rev_t.shape[0],
+                               _fori_body_fun,
+                               (rev_yt,
+                                rev_t,
+                                rev_tarray,
+                                rev_gi,
+                                vjp_y,
+                                vjp_t0,
+                                vjp_args,
+                                time_vjp_list))
+
+    time_vjp_list = jax.ops.index_update(result[-1], -1, result[-3])
+    vjp_times = np.hstack(time_vjp_list)[::-1]
+    return None, result[-4], vjp_times, unravel_args(result[-2])
+
+  return jax.jit(vjp_all)
+
+
+def test_grad_odeint():
```

Comment:
We should probably move these tests into the `tests/` directory like all the others, but that can wait until after we merge this.

Reply:
also: it would probably be cleaner to test against scipy's solvers.